core/src/main/scala/kafka/server/ReplicaFetcherThread.scala

```diff
@@ -254,15 +255,30 @@ class ReplicaFetcherThread(name: String,
       // We will not include a replica in the fetch request if it should be throttled.
       if (fetchState.isReadyForFetch && !shouldFollowerThrottle(quota, fetchState, topicPartition)) {
         try {
-          val logStartOffset = this.logStartOffset(topicPartition)
-          builder.add(topicPartition, new FetchRequest.PartitionData(
-            fetchState.fetchOffset, logStartOffset, fetchSize, Optional.of(fetchState.currentLeaderEpoch)))
+          // A fetch needs to be used to update the session only if:
+          //   1. fetch offset has changed
+          //   2. log start offset has changed
+          //   3. leader epoch has changed
+          //   4. the fetch was excluded before and now
+          if (fetchState.updateNeeded) {
+            val logStartOffset = this.logStartOffset(topicPartition)
+            builder.add(topicPartition, new FetchRequest.PartitionData(
+              fetchState.fetchOffset, logStartOffset, fetchSize, Optional.of(fetchState.currentLeaderEpoch)))
+          }
         } catch {
           case _: KafkaStorageException =>
             // The replica has already been marked offline due to log directory failure and the original failure should have already been logged.
             // This partition should be removed from ReplicaFetcherThread soon by ReplicaManager.handleLogDirFailure()
             partitionsWithError += topicPartition
         }
+      } else {
+        // Exclude this partition from the upcoming session;
+        // if we have to exclude this fetch because it is delayed, we need to update its state
+        // so that it would always be included when the delay has elapsed
+        builder.remove(topicPartition)
```

Comment:
Is it possible that we end up adding partitions to the "removed" list that we send in the incremental fetch request which were not in the previous request? Do we need to perform a contains check in here? https://github.com/apache/kafka/pull/7581/files#diff-38a1c48d28da7f7f671cef22421990e5R221

Reply:
1. Good catch, will do.
2. Hmm that's a very good question. Since all the access of the map is synchronized by the `partitionMapLock` I think we can maintain a removed partition synchronized by that map as well, and let the buildFetch call iterate over it and also clean it within the lock. That's not very elegant but I cannot come up with a better solution on top of my head now. WDYT @lbradstreet ?

Reply:
The only other solution I can really think of is to maintain a list of removed partitions, and clean them from the map as we do today? I'm not totally sure which feels cleaner.

Reply:
After some pondering on that I decided to add the `removedPartitions` in the ReplicaFetcher instead of PartitionStates, since it was covered in `inLock(partitionMapLock)` and is less  vulnerable to future changes.

Reply:
Sounds good. How is all of this integrating with the ReplicaAlterLogDirsThread implementation by the way? I'm guessing there's nothing to be done since there's no fetch request being built?